File: 0000001_twoSum/twoSum_test.go
package twosum_test

import (
	"testing"

	twosum "github.com/XOzymandiasz/leetcode-solutions/0000001_twoSum"
)

func TestTwoSum(t *testing.T) {
	tests := []struct {
		name   string
		nums   []int
		target int
	}{
		{"basic", []int{2, 7, 11, 15}, 9},
		{"unordered", []int{3, 2, 4}, 6},
		{"duplicates", []int{3, 3}, 6},
		{"negative", []int{7, 3, 4, -1}, 2},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			result := twosum.TwoSum(tc.nums, tc.target)

			if len(result) != 2 {
				t.Fatalf("expected 2 indices, got %v", result)
			}

			i, j := result[0], result[1]

			if i == j {
				t.Fatalf("indices must be different: %v", result)
			}

			if i < 0 || i >= len(tc.nums) || j < 0 || j >= len(tc.nums) {
				t.Fatalf("indices out of range: %v", result)
			}

			sum := tc.nums[i] + tc.nums[j]
			if sum != tc.target {
				t.Errorf(
					"nums[%d]=%d + nums[%d]=%d = %d, expected %d",
					i, tc.nums[i],
					j, tc.nums[j],
					sum,
					tc.target,
				)
			}
		})
	}
}

File: 0000001_twoSum/rust/tests/two_sum_test.rs
use rust::two_sum;

fn assert_valid(nums: &[i32], target: i32, result: &[i32]) {
    let i = result[0] as usize;
    let j = result[1] as usize;

    assert_ne!(i, j, "indices must be different: {:?}", result);

    let sum = nums[i] + nums[j];
    assert_eq!(
        sum, target,
        "nums[{}]={} + nums[{}]={} = {}, expected {}",
        i, nums[i], j, nums[j], sum, target
    );
}

#[test]
fn basic() {
    let nums = vec![2, 7, 11, 15];
    let result = two_sum(nums.clone(), 9);
    assert_valid(&nums, 9, &result);
}

#[test]
fn unordered() {
    let nums = vec![3, 2, 4];
    let result = two_sum(nums.clone(), 6);
    assert_valid(&nums, 6, &result);
}

#[test]
fn duplicates() {
    let nums = vec![3, 3];
    let result = two_sum(nums.clone(), 6);
    assert_valid(&nums, 6, &result);
}

#[test]
fn negative() {
    let nums = vec![7, 3, 4, -1];
    let result = two_sum(nums.clone(), 2);
    assert_valid(&nums, 2, &result);
}

#[test]
fn no_solution() {
    let nums = vec![1];
    let result = two_sum(nums.clone(), 2);
    assert_eq!(result, vec!{}, "[{}, {}], expected []", result[0], result[1]);
}